src/controller/update.ts

```diff
@@ -0,0 +1,235 @@
+/*
+This file contains functions that run on each frame.
+*/
+/*
+global
+G_utils_getSign
+G_model_isKeyDown
+G_model_actorSetVelocity
+G_model_actorSetPosition
+G_model_actorSetAcceleration
+G_model_actorSetFacing
+G_model_actorSetAnimState
+G_model_roomGetSizePx
+G_model_roomGetCollidableTiles
+G_model_tileIsFullyCollidable
+G_model_getFrameMultiplier
+G_utils_floorNearestMultiple
+G_utils_createRect
+G_utils_getCollisionsWithRect
+G_utils_rectToCollisionPoints
+G_KEY_LEFT
+G_KEY_RIGHT
+G_KEY_SPACE
+G_FACING_LEFT
+G_FACING_RIGHT
+G_ANIM_JUMPING
+G_ANIM_WALKING
+G_ANIM_DEFAULT
+G_COLLISION_TOP
+G_COLLISION_BOTTOM
+G_COLLISION_LEFT
+G_COLLISION_RIGHT
+*/
+
+const PLAYER_SPEED = 1.2; // the acceleration rate that the player moves left/right
+const PLAYER_JUMP_SPEED = 5.1; // the higher this value, the higher+faster the player jumps
+const AIR_CONTROL_DIVISOR = 30; // the higher this value, the less air control the player has.
+const MAX_SPEED_X = 1.2; // the maximum x speed that an actor can move
+const MAX_SPEED_Y = 3.2; // the maximum y speed that an actor can move
+const DECELERATION_RATE = 1.2; // the rate that actors decelerate (reduce velocity) per frame
+const GRAVITY_RATE = 0.2; // the rate that the y velocity is modified each frame (the higher, the faster the player moves down)
+
+// given an actor and a 'vx' or 'vy' reduce the velocity value one step closer to 0
+const decelerateVelocity = (actor: Actor, vel: 'vx' | 'vy') => {
+  const v = actor[vel];
+
+  if (v < 0) {
+    actor[vel] += DECELERATION_RATE;
+  } else if (v > 0) {
+    actor[vel] -= DECELERATION_RATE;
+  }
+
+  if (Math.abs(actor[vel]) <= DECELERATION_RATE) {
+    actor[vel] = 0;
+  }
+};
+
+const applyGravity = (actor: Actor) => {
+  const fm = G_model_getFrameMultiplier();
+  if (actor.isGround) {
+    actor.ay += 0.0001 * fm;
```

Comment:
This adds a sort of floatiness to when you move left or right off of a platform without jumping.

The game pushes the character down each frame as a simulation of gravity.  If the character is pushed into a platform, the character is moved back to the top of the platform and is marked as being on the ground.  If the character is in the air, then they are just accelerated downwards just like gravity would pull them down.

That works great.  Where it doesn't work how you'd expect is when a character walks Left or Right off of a platform.  When moving off the platform, the game accelerates the player downwards and doesn't hit a platform, so the actor is no longer on the ground.  We'd expect the actor to accelerate downwards, and that does happen, but because it is on the same frame as walking off the platform, the character moves downwards at a speed he accelerated downwards into the platform.  This results in a sort of 'fast fall' behavior that didn't feel consistent with most platformers.  So instead of applying all the gravity when a character is on the ground, a very small amount is applied so that when the character walks off of the platform, they don't immediately plummet at max speed downwards, but instead move at a small speed for the first frame, then move downwards faster.